File: Examples/ObjectiveC/BlockingGesturesDelegateExample.m
#import "BlockingGesturesDelegateExample.h"
@import Mapbox;

NSString *const MBXExampleBlockingGesturesDelegate = @"BlockingGesturesDelegateExample";

@interface BlockingGesturesDelegateExample () <MGLMapViewDelegate>
@property (nonatomic) MGLCoordinateBounds colorado;
@end

@implementation BlockingGesturesDelegateExample

- (void)viewDidLoad {
    [super viewDidLoad];
    
    MGLMapView *mapView = [[MGLMapView alloc] initWithFrame:self.view.bounds];
    mapView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    mapView.delegate = self;
    
    mapView.styleURL = [MGLStyle outdoorsStyleURLWithVersion:9];
    
    // Denver, Colorado
    CLLocationCoordinate2D center = CLLocationCoordinate2DMake(39.748947, -104.995882);
    
    // Starting point.
    [mapView setCenterCoordinate:center zoomLevel:10 direction:0 animated:NO];
    
    // Colorado's bounds
    CLLocationCoordinate2D ne = CLLocationCoordinate2DMake(40.989329, -102.062592);
    CLLocationCoordinate2D sw = CLLocationCoordinate2DMake(36.986207, -109.049896);
    self.colorado = MGLCoordinateBoundsMake(sw, ne);
    
    [self.view addSubview:mapView];
}

- (BOOL)mapView:(MGLMapView *)mapView shouldChangeFromCamera:(MGLMapCamera *)oldCamera toCamera:(MGLMapCamera *)newCamera
{
    // Get current coordinates
    MGLCoordinateBounds visibleCoordinateBounds = mapView.visibleCoordinateBounds;
    CLLocationCoordinate2D newCameraCenter = newCamera.centerCoordinate;
    CLLocationCoordinate2D oldCameraCenter = oldCamera.centerCoordinate;
    
    // Get the offset from old camera center to current visible map bounds
    CGFloat neLatitudeOffset = visibleCoordinateBounds.ne.latitude - oldCameraCenter.latitude;
    CGFloat neLongitudeOffset = visibleCoordinateBounds.ne.longitude - oldCameraCenter.longitude;
    CGFloat swLatitudeOffset = visibleCoordinateBounds.sw.latitude - oldCameraCenter.latitude;
    CGFloat swLongitudeOffset = visibleCoordinateBounds.sw.longitude - oldCameraCenter.longitude;
    
    // Update the boundaries with new camera center + boundary offset
    CLLocationCoordinate2D newNE = CLLocationCoordinate2DMake(neLatitudeOffset + newCameraCenter.latitude, neLongitudeOffset + newCameraCenter.longitude);
    
    CLLocationCoordinate2D newSW = CLLocationCoordinate2DMake(swLatitudeOffset + newCameraCenter.latitude, swLongitudeOffset + newCameraCenter.longitude);
    MGLCoordinateBounds newBounds = MGLCoordinateBoundsMake(newSW, newNE);
    
    // Test if the new camera center point and boundaries are inside colorado
    BOOL inside = MGLCoordinateInCoordinateBounds(newCameraCenter, self.colorado);
    BOOL intersects = MGLCoordinateInCoordinateBounds(newBounds.ne, self.colorado) && MGLCoordinateInCoordinateBounds(newBounds.sw, self.colorado);
    
    return inside && intersects;

}

@end

File: Examples/Swift/BlockingGesturesDelegateExample.swift
import Mapbox

@objc(BlockingGesturesDelegateExample_Swift)

class BlockingGesturesDelegateExample_Swift: UIViewController, MGLMapViewDelegate {

    private var colorado: MGLCoordinateBounds!

    override func viewDidLoad() {
        super.viewDidLoad()

        let mapView = MGLMapView(frame: view.bounds)
        mapView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        mapView.delegate = self

        // Denver, Colorado
        let center = CLLocationCoordinate2D(latitude: 39.748947, longitude: -104.995882)

        // Starting point
        mapView.setCenter(center, zoomLevel: 10, direction: 0, animated: false)

        // Colorado’s bounds
        let northeast = CLLocationCoordinate2D(latitude: 40.989329, longitude: -102.062592)
        let southwest = CLLocationCoordinate2D(latitude: 36.986207, longitude: -109.049896)
        colorado = MGLCoordinateBounds(sw: southwest, ne: northeast)

        view.addSubview(mapView)
    }

    // This example uses Colorado’s boundaries to restrict the camera movement.
    func mapView(_ mapView: MGLMapView, shouldChangeFrom oldCamera: MGLMapCamera, to newCamera: MGLMapCamera) -> Bool {

        // Get the current camera to restore it after.
        let currentCamera = mapView.camera

        // From the new camera obtain the center to test if it’s inside the boundaries.
        let newCameraCenter = newCamera.centerCoordinate

        // Set the map’s visible bounds to newCamera.
        mapView.camera = newCamera
        let newVisibleCoordinates = mapView.visibleCoordinateBounds

        // Revert the camera.
        mapView.camera = currentCamera

        // Test if the newCameraCenter and newVisibleCoordinates are inside self.colorado.
        let inside = MGLCoordinateInCoordinateBounds(newCameraCenter, self.colorado)
        let intersects = MGLCoordinateInCoordinateBounds(newVisibleCoordinates.ne, self.colorado) && MGLCoordinateInCoordinateBounds(newVisibleCoordinates.sw, self.colorado)

        return inside && intersects
    }
}
